File: src/main/java/net/bbelovic/adventofcode/day2/Day2PuzzlePart2.java
package net.bbelovic.adventofcode.day2;

import net.bbelovic.adventofcode.Puzzle;

class Day2PuzzlePart2 implements Puzzle<String, Long> {
    @Override
    public Long solve(String input) {
        long[] dimensions = parseDimensions(input);
        return computeRibbon(dimensions);
    }

    private long computeRibbon(long [] dimensions) {
        long l = dimensions[0];
        long w = dimensions[1];
        long h = dimensions[2];
        long first = Math.min(2 * (l + w), 2*(w+h));
        long second = Math.min(2 * (w + h), 2 * (h + l));
        long perimeter = Math.min(first, second);
        return perimeter + l*w*h;
    }

    private long [] parseDimensions(String input) {
        String[] split = input.split("x");
        return new long []{
                Long.valueOf(split[0]),
                Long.valueOf(split[1]),
                Long.valueOf(split[2]),
        };
    }
}

File: aoc-2018/src/main/kotlin/net/bbelovic/adventofcode/year2018/day2/Day2PuzzlePart2.kt
package net.bbelovic.adventofcode.year2018.day2

import net.bbelovic.adventofcode.Puzzle
import java.lang.StringBuilder
import java.util.ArrayList

class Day2PuzzlePart2 : Puzzle<List<String>, String> {
    override fun solve(input: List<String>): String {
        for (each in input) {
            val copy = ArrayList(input - each)
            for (s in copy) {
                if (!hasSingleDifference(each, s)) {
                    continue
                } else {
                    return findCommonLetters(each, s)
                }
            }
        }
        throw IllegalStateException("Expected exactly one pair of IDs that differ in one letter.")
    }

    private fun hasSingleDifference(first: String, second: String): Boolean {
        var diffs = 0
        for (i in 0 until first.length) {
            if (first[i] != second[i]) {
                ++diffs
            }
            if (diffs >= 2) break
        }
        return diffs == 1
    }

    private fun findCommonLetters(first: String, second: String): String {
        val result = StringBuilder()
        for (i in 0 until first.length) {
            if (first[i] == second[i]) result += first[i]
        }
        return result.toString()
    }

    operator fun StringBuilder.plusAssign(c: Char) {
        this.append(c)
    }
}